doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,247 @@
+      void onSampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived2);
+      }
+
+      myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+      ```
+ - One can attach at most one callback to a specific event of an object.
+    - Attaching a callback to an event where a callback has been already attached overrides
+      the existing callback with the new one.
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished.
+    - Calling detach means that after that call the callback is no longer attached
+      even when it was not attached in the first place. Therefore it will always succeed.
+
+ - When the ReactAl goes out of scope it should detach itself from every class 
+     to which it was attached via a provided callback (like in the WaitSet).
+
+ - When the class which is attached to the ReactAl goes out of scope it should 
+    detach itself from the ReactAl via a provided callback (like in the WaitSet).
+
+### Usage Code Example
+```cpp
+ReactAl myCallbackReactal;
+iox::popo::UntypedSubscriber mySubscriber;
+
+void onSampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+  subscriber.take().and_then([&](auto & sample){
+    std::cout << "received " << std::hex << sample->payload() << std::endl;
+  });
+}
+
+void onWaitForSubscription(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "subscribed!\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, onSampleReceived);
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED, onWaitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED);
+}
+
+void onWaitForUnsubscribe(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "unsubscribed from publisher\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED, onWaitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED);
+}
+
+int main() {
+  myCallbackReactal.attachEvent(mySubscriber, iox::popo::SubscriberEvent::SUBSCRIBED, onWaitForSubscription);
+}
+```
+
+### Solution
+#### Condition Variable
```

Comment:
I always think the building block for this should be a Semaphore (with the ability to exist in shared memory if needed). It is not actually a condition variable in my book.